services/cluster/mutex.go

```diff
@@ -0,0 +1,201 @@
+	ttl = time.Second * 15
+
+	// refreshInterval is the interval on which the mutex will be refreshed when locked
+	refreshInterval = ttl / 2
+
+	// minWaitInterval is the minimum amount of time to wait between locking attempts
+	minWaitInterval = 1 * time.Second
+
+	// maxWaitInterval is the maximum amount of time to wait between locking attempts
+	maxWaitInterval = 5 * time.Minute
+
+	// pollWaitInterval is the usual time to wait between unsuccessful locking attempts
+	pollWaitInterval = 1 * time.Second
+
+	// jitterWaitInterval is the amount of jitter to add when waiting to avoid thundering herds
+	jitterWaitInterval = minWaitInterval / 2
+)
+
+// StoreAPI is the API interface required to manage mutexes.
+type StoreAPI interface {
+	CompareAndSet(kv *model.AtomicKeyValue, oldValue []byte) (bool, *model.AppError)
+	CompareAndDelete(key string, oldValue []byte) (bool, *model.AppError)
+}
+
+// DBMutex mutex usable across cluster that uses database as a backend
+type DBMutex struct {
+	store StoreAPI
+	key   string
+
+	// lock guards the variables used to manage the refresh task, and is not itself related to
+	// the cluster-wide lock.
+	lock        sync.Mutex
+	stopRefresh chan bool
+	refreshDone chan bool
+}
+
+// NewDBMutex creates a mutex with the given name.
+func NewDBMutex(key string, storeAPI StoreAPI) *DBMutex {
+	key = mutexPrefix + key
+
+	return &DBMutex{
+		store: storeAPI,
+		key:   key,
+	}
+}
+
+// lock makes a single attempt to atomically lock the mutex, returning true only if successful.
+func (m *DBMutex) tryLock() (bool, error) {
+	ok, err := m.store.CompareAndSet(&model.AtomicKeyValue{
+		Key:          m.key,
+		Value:        []byte{1},
+		ExpireInSecs: int64(ttl / time.Second),
+	}, nil)
+	if err != nil {
+		return false, errors.Wrap(err, "failed to set mutex kv")
+	}
+
+	return ok, nil
+}
+
+// refreshLock rewrites the lock key value with a new expiry, returning true only if successful.
+//
+// Only call this while holding the lock.
+func (m *DBMutex) refreshLock() error {
+	ok, err := m.store.CompareAndSet(&model.AtomicKeyValue{
+		Key:          m.key,
+		Value:        nil,
+		ExpireInSecs: int64(ttl / time.Second),
+	}, []byte{1})
+	if err != nil {
+		return errors.Wrap(err, "failed to refresh mutex kv")
+	} else if !ok {
+		return errors.New("unexpectedly failed to refresh mutex kv")
+	}
+
+	return nil
+}
+
+// Lock locks m. If the mutex is already locked by any plugin instance, including the current one,
+// the calling goroutine blocks until the mutex can be locked.
+func (m *DBMutex) Lock() {
+	var waitInterval time.Duration
+
+	for {
+		time.Sleep(waitInterval)
+
+		locked, err := m.tryLock()
+		if err != nil {
+			mlog.Error("failed to lock mutex", mlog.Err(err), mlog.String("lock_key", m.key))
+			waitInterval = nextWaitInterval(waitInterval, err)
+			continue
+		} else if !locked {
+			waitInterval = nextWaitInterval(waitInterval, err)
+			continue
+		}
+
```

Comment:
This part can be refactored into a function along with the infinite loop. Because the moment we enter the loop, we just return after we are done. We can call the function `waitUntilLockAcquired`.

Reply:
done